torch/sparse/semi_structured.py

```diff
@@ -0,0 +1,318 @@
+            return SparseSemiStructuredTensor(
+                args[0].shape,
+                args[0].compressed_tensor,
+                args[0].cslt,
+                not args[0].transposed,
+            )
+
+        # handle addmm
+        if func is torch.ops.aten.addmm.default:
+            bias, input_A, input_B = args
+
+            # Currently, we only support the first matrix being sparse for addmm/mm in cuSPARSELT and CUTLASS.
+            # CUTLASS only supports the first input to be sparse for a given matmul.
+            # cuSPARSELt does not have this limitation, although it appears that it uses CUTLASS under the hood,
+            # since it will be slower to have the second matrix be sparse.
+            # It may be using the same transpose trick we are using below
+            if (
+                isinstance(input_A, cls)
+                and not input_A.transposed
+            ):
+                return input_A.cslt.addmm(
+                    input_B, bias, not input_B.is_contiguous(), False
+                )  # type: ignore[attr-defined]
+
+            # Although we only support the first matrix being sparse, we can support the second matrix being sparse.
+            # We do this by taking advantage of some transpose properties:
+            # F.linear(x) = addmm(bias, input, weight.t()) = b + xW' = (b + xW')''
+            #        = (W''x' + b')' = (Wx' + b')' = W.cslt.addmm(input, ).T
+            elif isinstance(input_B, cls) and input_B.transposed:
+                res = input_B.t().cslt.addmm(input_A.T, bias, True, cls._fuse_transpose)  # type: ignore[attr-defined]
+                return res if cls._fuse_transpose else res.T
+
+            raise NotImplementedError(
+                (
+                    f"func: {func} is currently not supported for ",
+                    f"opA: {input_A.transposed} A: {input_A}",
+                    f"opB: {input_B.transposed} B {input_B}",
+                )
+            )
+
+        if func is torch.ops.aten.mm.default:
+            input_A, input_B = args
+
+            if (
+                isinstance(input_A, cls)
+                and not input_A.transposed
+            ):
+                return input_A.cslt.mm(input_B, not input_B.is_contiguous(), False)  # type: ignore[attr-defined]
+
+            elif isinstance(input_B, cls) and input_B.transposed:
+                res = input_B.t().cslt.mm(input_A.T, True, cls._fuse_transpose)  # type: ignore[attr-defined]
+                return res if cls._fuse_transpose else res.T
+
+            raise NotImplementedError(
+                (
+                    f"func: {func} is currently not supported for ",
+                    f"opA: {input_A.transposed} A: {input_A}",
+                    f"opB: {input_B.transposed} B {input_B}",
+                )
+            )
+
+        # When torch is run with inference mode, it looks like pytorch does some merging in order to make linear faster.
```

Comment:
Actually it's slightly different than that: we do not do autograd decompositions before getting here in inference mode. Hence this behavior. 
I'm sure we could do something better here...

Reply:
Do you mean a better way to handle autograd decompositions or a better way to handle inference_mode?

Reply:
A better way for you to apply the autograd decomposition even in inference mode as it seems to be what you want.
cc @bdhirsh FYI